src/proxy.js

```diff
@@ -0,0 +1,278 @@
+            groups,
+            country,
+            password = process.env[ENV_VARS.PROXY_PASSWORD],
+            hostname = process.env[ENV_VARS.PROXY_HOSTNAME] || LOCAL_ENV_VARS[ENV_VARS.PROXY_HOSTNAME],
+            port = parseInt(process.env[ENV_VARS.PROXY_PORT] || LOCAL_ENV_VARS[ENV_VARS.PROXY_PORT], 10),
+        } = options;
+
+        // Validation
+        checkParamOrThrow(groups, 'opts.groups', 'Maybe [String]');
+        checkParamOrThrow(country, 'opts.country', 'Maybe String');
+        checkParamOrThrow(password, 'opts.password', 'Maybe String');
+        checkParamOrThrow(hostname, 'opts.hostname', 'String', this._getMissingParamErrorMgs('hostname', ENV_VARS.PROXY_HOSTNAME));
+        checkParamOrThrow(port, 'opts.port', 'Number', this._getMissingParamErrorMgs('port', ENV_VARS.PROXY_PORT));
+        this._validateArgumentsStructure(groups, country);
+
+        this.groups = groups;
+        this.country = country;
+        this.password = password;
+        this.hostname = hostname;
+        this.port = port;
+        this.availableGroups = [];
+    }
+
+    /**
+     * Loads proxy password if token is provided and checks access to apify proxy and privided proxy groups.
+     * This function must be called before you can start using the instance in a meaningful way.
+     *
+     * @returns {Promise<void>}
+     */
+    async initialize() {
+        // Set proxy password via API if token is provided
+        await this._setPasswordIfToken();
+
+        // Check if user has access to apify proxy and selected proxy groups
+        await this._checkAccess();
+
+        // TODO: Validate proxyUrl each of custom proxies
+    }
+
+
+    /**
+     * Gets information about proxy and its configuration.
+     * @param {string} sessionId
+     * Apify Proxy session identifier to be used with requests.
+     * All HTTP requests going through the proxy with the same session identifier
+     * will use the same target proxy server (i.e. the same IP address).
+     * The identifier can only contain the following characters: `0-9`, `a-z`, `A-Z`, `"."`, `"_"` and `"~"`.
+     * @return {object} represents information about proxy.
+     */
+    getInfo(sessionId) {
+        const { groups, country, password, port, hostname } = this;
+        const url = this.getUrl(sessionId);
+
+        return {
+            sessionId,
+            url,
+            groups,
+            country,
+            password,
+            hostname,
+            port,
+        };
+    }
+
+    /**
+     * Returns proxy url.
+     * @param {string} sessionId
+     * Apify Proxy session identifier to be used with requests.
+     * All HTTP requests going through the proxy with the same session identifier
+     * will use the same target proxy server (i.e. the same IP address).
+     * The identifier can only contain the following characters: `0-9`, `a-z`, `A-Z`, `"."`, `"_"` and `"~"`.
+     * @return {string} - the proxy url
+     */
+    getUrl(sessionId) {
```

Comment:
How about calling this `getProxyUrl` to make it super-clear? Everywhere we always talk about "proxy url", so having just "getUrl" might be bit ambiguous.

Reply:
I don't like `proxyConfiguration.getProxyUrl()` for the double proxy, but have no strong feelings about this. Also, for consistency, the constructor options would have to be 

```js
await Apify.createProxyConfiguration({
    proxyPassword: 'sfsdfs',
    proxyGroups: [ ... ],
    ...
})
```

But, thinking about it now, we might actually want to name it `createUrl()`, because it actually creates it using the `sessionId`. It's not there to begin with.

Reply:
Or maybe not. `getThing(thingId)` is quite valid. Even though here, the IDs are of something else. `getUrl(sessionId)`... Gah, the naming :D

Reply:
Yeah, naming is great :) For `Apify.createProxyConfiguration()`, having there `proxyPassword` etc would look quite terrible. Regarding `getUrl` / `getProxyUrl` / `generateUrl` / `generateProxyUrl` / ..., it was just my feeling, I'll leave it with you guys...